### ISSUE TITLE
Questions on callbacks

### ISSUE THREAD

**Issue by Indigo744** [neveldo/jQuery-Mapael#179]
Hey,

I was wondering regarding some callbacks we provide: `afterInit`, `afterUpdate`, `afterShowRange`
What is the purpose of using an "after" callback, instead of simply writing a function after the call? Since JS is sequential, this will work the same.

I can understand the need for `afterInit` and `afterUpdate`, they provide useful parameters (paper, areas...), even though these parameters are accessible through the Mapael object.

In any way, `afterShowRange` is useless in this current state.

I was thinking, maybe the callbacks should fire only when all animations (if there are animation of course) are finished? That way, the callback will have a more useful purpose.

Right now, the user doesn't have any information regarding when the animation are done. If a user wants to do something else, it would have to use a `setTimeout()` in order to wait for any animation to finish.

Actually, I was also thinking that any trigger function could return a promise (See the [deferred](https://api.jquery.com/category/deferred-object/) object). That way, the user could write:
```javascript
$(".mapael").trigger("update", opt).done(function(){
    // ... special processing executed only after all animations are done
});
```

What is your opinion on this?

**Comment by Indigo744** [comment]
Also, the Zoom trigger could use an `afterZoom` callback fired only when the zoom is complete.

**Comment by neveldo** [comment]
Hello,

The main goal of these events is indeed to provide some hooks to the user for several steps during the execution of mapael, and to provide internal objects / values to play with them.

As all the parameters are now accessible through the Mapael object, some of these events may be useless now (except for keeping the backward compatibility of couse). 

As you mentionned, we could enhance these events in order to take into account the timeout option by triggering through a setTimeout() call, I think its a great idea !

**Comment by Indigo744** [comment]
**Update for the deferred part:**
Trying to do like in the example (i.e. calling `done()` after `trigger()`) will not work as the `trigger('update')` function always returns the jQuery objet itself (and not what the 'update' function actually returns).
Solution is to use the `triggerHandler('update')` function, which will return the actual 'update' function return value.
See this SO thread: http://stackoverflow.com/questions/25348842/deferred-jquery-event-trigger

Hence, I don't think it's really interesting to use a deferred object. Let's stay simple.

`afterZoom` callback may be nice though.

Added to TODO list.

**Comment by neveldo** [comment]
Hello,

Thank you for having checked for the use of a deferred object. Regarding 'afterZoom', actually, it already exists.

In any case, some of the triggers don't take into account the timeout (in the update or the zoom events for instance) so maybe the triggering of these callbacks could be enhanced a little.

**Comment by Indigo744** [comment]
Right, I forgot about this `afterZoom` callback that can be set globally!
